### PR TITLE
Finished up on shooter structure

### PR DESCRIPTION
**Pull request by rohitalamgari**

### CODE REVIEW

File: Competition/src/main/cpp/subsystems/Shooter.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+void Shooter::assignOutputs()
+{   
+    state.turretTarget = 0;
+
+    //DISABLED
+    if(state.turretState == TurretState::TURRET_DISABLE){
+        state.turretTarget = 0;
+    }//HOME
+    else if(state.turretState == TurretState::TURRET_HOME){
+        // @TODO Motion Profile to turret home position
```
Comment (mray190) [inline Competition/src/main/cpp/subsystems/Shooter.cpp:135]: Need motion profile code in place. PID constants can be set to 0 for now, and target positions can be randomly assigned as long as a constant is created and can be modified

File: Competition/src/main/cpp/subsystems/Shooter.cpp [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        // Stop deadband
+        else if (std::abs(state.turretTarget) < ShooterConstants::pSoftDeadband) {
+            int direction = 1;
+            if (state.turretTarget < 0) direction = -1;
+            state.turretTarget = ShooterConstants::pSoftDeadband * direction;
+        }
+    }//DEFAULT
+    else if (state.turretState == TurretState::TURRET_DEFAULT){
+        //Odometry tracking
+        // @TODO Motion Profile to target turret position
```
Comment (mray190) [inline Competition/src/main/cpp/subsystems/Shooter.cpp:155]: Same as before - need motion profile code and constants

File: Competition/src/main/cpp/subsystems/Shooter.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    turret.Set(state.turretTarget);
+
+    if(state.flywheelState == FlywheelState::FLYWHEEL_DISABLE){
+        state.flywheelTarget = 0;
+    }
+    else if(state.flywheelTarget == FlywheelState::FLYWHEEL_PRIME){
+        //math conditions, leaving 1 for now
+        // @TODO use dashboard value
```
Comment (mray190) [inline Competition/src/main/cpp/subsystems/Shooter.cpp:166]: Value should be changed via a dashboard item

File: Competition/src/main/cpp/subsystems/Shooter.cpp [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        state.flywheelTarget = 1;
+    }
+    else if (state.flywheelTarget == FlywheelState::FLYWHEEL_DEFAULT){
+        // @TODO use dashboard value
```
Comment (mray190) [inline Competition/src/main/cpp/subsystems/Shooter.cpp:170]: Value should be changed via a dashboard item

File: Competition/src/main/cpp/subsystems/Shooter.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        state.flywheelTarget = 0.5;
+    }
+    
+    flywheel_lead.Set(state.flywheelTarget);
+
+    if(state.hoodState == HoodState::HOOD_DISABLE){
+        state.hoodTarget = 0;
+    }
+    else if(state.hoodState == HoodState::HOOD_PRIME){
+        // @TODO Motion profile to target hood position
```
Comment (mray190) [inline Competition/src/main/cpp/subsystems/Shooter.cpp:180]: Need motion profile as before

**Review by mray190 — CHANGES_REQUESTED** [review]
Needs changes. See line by line comments

File: Competition/src/main/cpp/subsystems/Shooter.cpp (hunk 3 above)
Comment (janhelgeson) [reply Competition/src/main/cpp/subsystems/Shooter.cpp]: Additionally need to use falcon control mode for velocity

**Comment by janhelgeson** [comment]
Need shooter flywheel stuff to be falcon instead of rev

File: Competition/src/main/cpp/subsystems/Shooter.cpp [hunk 6]
```diff
@@ -34,21 +34,48 @@ void Shooter::init()
     limeTable = nt::NetworkTableInstance::GetDefault().GetTable("limelight");
     initTable("Shooter");
     table->PutBoolean("Home Turret", false);
+    table->PutNumber("Flywheel Primed Value", 1);
+    table->PutNumber("Flywheel Default Value", 0.5);
+    table->PutNumber("Hood Top Position", 200);
```
Comment (mray190) [inline Competition/src/main/cpp/subsystems/Shooter.cpp:39]: Put default in constants

File: Competition/src/main/cpp/subsystems/Shooter.cpp [hunk 7]
```diff
@@ -140,94 +155,116 @@ void Shooter::assessInputs()
     else if (state.backButton){
         state.flywheelState = FlywheelState::FLYWHEEL_DEFAULT;
     }
+
+    if (state.rightBumper){
```
Comment (mray190) [inline Competition/src/main/cpp/subsystems/Shooter.cpp:159]: Use ternary operator

File: Competition/src/main/cpp/subsystems/Shooter.cpp [hunk 8]
```diff
@@ -140,94 +155,116 @@ void Shooter::assessInputs()
...
     if(table->GetBoolean("Home Turret", false)){
         state.turretState = TurretState::TURRET_HOME;
     }
+
+    //slider
+    state.flywheelLow = table->GetNumber("Flywheel Default Value", .5);
+    state.flywheelHigh = table->GetNumber("Flywheel Primed Value", 1);
+
+    state.hoodLow = table->GetNumber("Hood low position", 0);
+    state.hoodHigh = table->GetNumber("Hood high position", 2000);
```
Comment (mray190) [inline Competition/src/main/cpp/subsystems/Shooter.cpp:178]: Default values should match the constants value (See earlier comment)

File: Competition/src/main/cpp/subsystems/Shooter.cpp [hunk 9]
```diff
@@ -140,94 +155,116 @@ void Shooter::assessInputs()
...
-        state.turretPositionSetpoint = ShooterConstants::homeFrontPosition;
+        state.turretTarget = ShooterConstants::homeFrontPosition;
         useSmartMotion = true;
     }
 
     //PRIMED
     else if (state.turretState == TurretState::TURRET_PRIME){
-        //limelight
-        useSmartMotion = true;
+        float tx = limeTable->GetNumber("tx", 0.0);
```
Comment (mray190) [inline Competition/src/main/cpp/subsystems/Shooter.cpp:216]: Need logic to turn on and off the LED

File: Competition/src/main/include/subsystems/Shooter.h [hunk 10]
```diff
@@ -48,6 +48,10 @@ class Shooter : public ValorSubsystem
 
      double getTargetTics(double, double, double, double, double, double, double);
      double convertTargetTics(double, double, double);
+
+     //@TODO
```
Comment (mray190) [inline Competition/src/main/include/subsystems/Shooter.h:52]: Remove todo

**Review by mray190 — CHANGES_REQUESTED** [review]
See comments

**Pull request closed without merging**